WIP: threadpool: pluggable threadpool

davisjam opened the pull request:
Edit: Since this discussion is fairly long and the PR is at the bottom, here are some shortcuts for newcomers. A working implementation is described [below](https://github.com/libuv/libuv/pull/1726#issuecomment-416711557). See also #1959.

-----------------

WIP For #267.

@bnoordhuis 
@saghul 
@sam-github (cf. #1528)

davisjam commented:
@sam-github I'm trying to accommodate #1528 with a `uv_threadpool_monitor_t` type.

davisjam commented:
Perhaps it will be more productive to discuss based on what's in the docs (58cba8c). Should be clearer than my half-baked declarations.

gireeshpunathil commented:
few open questions:

 - what are the key attributes that differentiates one TP from another? # of resident threads, CPU affinity, priority, work scheduling / work queue model? what are user-customizable and what are not? Looking at the API `uv_threadpool_create` , I am able to see ony thread count. In other words, how do we `qualify` a TP for a CPU bound task?
 - will each TP be bound to its own loop, or to a common loop? 
 - will the internal complexity of concurrency be now handled by libuv consumers as opposed to libuv itself? In other words, will the main thread's execution behave differently (in terms of performance) in purview of the presence of a customized and tuned threadpool?

Thanks!

saghul commented:
IMHO this would be better targeted at master, and thanks for getting the ball rolling again!

bnoordhuis commented:
As long as it doesn't break API/ABI, I'm good with targeting v1.x.

davisjam commented:
@gireeshpunathil 

> what are the key attributes that differentiates one TP from another?

At the moment, there are two differentiating aspects between all threadpools:
- the *loop* to which each is bound (completed work will be placed on this loop's done queue to run the `after_work_cb`s)
- the *size* (number of threads)

In addition, custom threadpools also have their own queue and cancel functions so they can do things like priority queuing.

> how do we qualify a TP for a CPU bound task?

You create a TP and only assign CPU-bound work to it, like this:

```C
// Initialization
uv_threadpool_t* cpu_tp = uv_create_threadpool(loop, 4, ...);
...
// Use later on
uv_queue_work(cpu_tp, work);
```

> will each TP be bound to its own loop, or to a common loop?

Each `uv_loop_t` will have a single unified queue of done tasks, and zero or more TPs can be bound to a loop and have their done tasks placed in this queue.

> will the internal complexity of concurrency be now handled by libuv consumers as opposed to libuv itself?

Only if the libuv consumers want to be fancy. Users can create as many "default" TPs as they want, each of which will use the current FIFO model. Users interested in fancier concurrency can implement a custom TP.

> CPU affinity, priority, work scheduling / work queue model?

Each in turn:
- *CPU affinity*: I like this idea -- allowing creator to request a CPU mask for the threads. I'll introduce a `uv_threadpool_options_t` that can encode this.
- *priority*: User can implement this with a custom TP.
- work scheduling/work queue model: User can implement this with a custom TP (though I need to explore appropriate CBs for this).

davisjam commented:
> As long as it doesn't break API/ABI, I'm good with targeting v1.x.

It might, but I'll keep it at v1.x for now.

davisjam commented:
7b4ea9c states:

1. The external API I'm proposing.
2. The internal CBs that a TP implementation has to implement in order to support this external API.

davisjam commented:
2dbfca9 tweaks the internal CBs for a TP.

davisjam commented:
Circling towards an API. Types and declarations in 9d63121.

davisjam commented:
Ben, thanks for the feedback.

I'm going to respond with what I think is essentially the same point made by @saghul [here](https://github.com/libuv/libuv/issues/267#issuecomment-78951982), but...

> Node.js or an add-on calls uv_queue_work() or another API that uses the thread pool

If the add-on gives Node a black-box work item via `uv_queue_work`, I don't see how Node can make effective scheduling decisions about it.

> It would make experimentation with different thread pool (or pools) configurations much easier for Node.js because all the interesting logic shifts to the Node.js side.

It sounds like you're talking about Node re-implementing the libuv threadpool. Node can do this *right now* if it wants to, with `uv_thread_t`'s and a `uv_async_t` to fire a queue of done callbacks. No need for libuv support. The libuv TP just uses lower-level public libuv APIs anyway.

The contribution libuv makes in the TP space is (1) managing the threads, (2) offering a simple FIFO TP implementation out of the box, and (3) defining a good API for TPs (e.g. with support for monitoring its state). *I think the TP *thread* code should be implemented in libuv, while the TP queueing/monitoring code can be handled by the embedding application if they want.* That's what my proposed API enables via the "custom" TP.

>  That's not possible when add-ons have their own threadpools or Node.js has no insight into what goes in and out.

I agree that Node add-ons should probably not be creating their own threadpools, and that was not my intent with this API.  I think Node as the embedding application can expose its own set of TP APIs, and the Node docs can then strongly advise add-ons not to use the libuv TP API. I think there are other libuv features that add-ons should use the Node equivalents of rather than talking to libuv directly.

Strawman: Node could create one standard FIFO TP for CPU-bound work, using the proposed `uv_threadpool_create_default` API, and expose it to add-ons via

```C++
node_queue_CPU_work(node_work_t* w);
```

An analogous API for I/O-bound work could route to the default libuv TP which would also handle libuv FS and DNS requests.

```C++
node_queue_IO_work(node_work_t* w);
```

> enable offloading scheduling to the embedding application

The Node TPs I'm imagining do allow the embedding application to control scheduling. Look at these CBs for a "custom" TP:

```
uv_threadpool_queue_work_cb:   Place work on the TP queue
             Could be invoked by any thread via uv_threadpool_queue_work
uv_threadpool_dequeue_work_cb: Get work from the TP queue (non-blocking)
             Called by threadpool workers.
```

If the application controls the queue order, it effectively controls the schedule order. Though it won't control precisely when tasks are scheduled, it does controls the order in which they are scheduled, which is about as much control as I think it needs.

One thing this API does *not* support is a dynamically-sized TP. I'm open to adding APIs that add or subtract threads in a TP.

davisjam commented:
The weekend is coming, when I usually do some work on this. @bnoordhuis any thoughts?

bnoordhuis commented:
Sorry about that.  Had a half-finished reply sitting for a few days.

> If the add-on gives Node a black-box work item via uv_queue_work, I don't see how Node can make effective scheduling decisions about it.

I'd call it gray.  Node can still make some scheduling decisions based on whether the work item comes from an add-on or node.js itself.

My intuition is that we'd offload add-on work items to the same thread pool that we'd use for crypto tasks (i.e. long-running, CPU-intensive); we could then iterate on an API that lets add-ons provide hints to node.js, sorta like your strawman.

> It sounds like you're talking about Node re-implementing the libuv threadpool. Node can do this right now if it wants to, with uv_thread_t's and a uv_async_t to fire a queue of done callbacks.

That's right, and we already do to some extent.  The reason I propose this API is that if node.js stands to benefit from it, then it will probably help other projects as well.  Not having to reinvent the wheel and all that.

> One thing this API does not support is a dynamically-sized TP.

That's one reason for wanting to push this to the embedder, because it would let us experiment with that more easily - and also because I suspect there's no one-size-fits-all solution.  libeio (when we were still using that) had a minthreads/maxthreads api and it only worked and performed so-so.

davisjam commented:
OK, I think I better understand where you're coming from now.

> I'd call it gray

Could be improved if the we support hints of what the work will do -- at the most basic, something like IO vs. CPU and small vs. large.

To my mind your proposal provides the maximal control to the embedder. But if the embedder just wants multiple FIFO-style TPs then this is overkill, and they will have to implement my version of the "pluggable libuv TP" API themselves. Seems like a waste.

I'm open to implementing both. My version should be a not-too-bad refactoring of the existing TP, and your version basically just requires a new API.

bnoordhuis commented:
Yours and mine might not be incompatible.  The idea is to provide a highly flexible API for embedders that want that kind of control, and implement something on top of that offers a good out-of-the-box experience for less picky embedders.

davisjam commented:
Agreed. I'll work on an API.

davisjam commented:
@bnoordhuis: I think an implementation of my proposal will be good for libuv-centric applications, and will be helpful building block for embedders interested in your proposal. I think they're both useful in their own ways.

My new design is in docs/src/threadpool.rst (surprise, in Markdown format!).

davisjam commented:
I am working on this again.

@bnoordhuis: Did you ever follow through on your plan to make a meta-issue to track the state of the threadpool proposals? I've just gone through all of the issues I could find that referenced the threadpool, proposed designs, or requested features.

davisjam commented:
@addaleax I am thinking about reverting your #1845 as part of this PR. That logic can go in the TP that Node.js will plug in. What do you think?

davisjam commented:
@bnoordhuis @saghul See proposed API in 7dcc0ce.

addaleax commented:
@davisjam I assume libuv would keep going with the current implementation as its default, right? In that case I’m not sure how I feel about that – 1845 fixed a real bug, imo…

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -52,15 +96,59 @@ Public members
+
+    Boolean.
+    If truth-y, it is safe to abort this work while it is being handled
+    by a thread.
```
Suggestion: s/truth-y/nonzero/

It might be good to specify what `abort` means? As I understand it, currently libuv only cancels work requests that are not yet executing… does this map to that 1:1?

davisjam replied:
> currently libuv only cancels work requests that are not yet executing

Yes. This has proved unfortunate in the case of DNS requests + shutdown of Node.js. See for example [here](https://github.com/libuv/libuv/issues/203#issuecomment-74080305) and [here](https://github.com/nodejs/node/issues/795#issuecomment-74063491).

My intention is to permit an executor to cancel work requests that have been assigned to threads, if they indicate that they are cancelable and if the executor so desires.

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -52,15 +96,59 @@ Public members
+            UV_WORK_USER_IO,
+            UV_WORK_USER_CPU,
+            UV_WORK_PRIVATE,
+            UV_WORK_TYPE_MAX
```
This makes adding a new entry a breaking change, right? I think that could be avoided by making `UV_WORK_TYPE_MAX` the return value of a function.

davisjam replied:
I think you're right ABI-wise but am not certain. Can also be addressed by setting UV_WORK_TYPE_MAX=255.

davisjam commented:
@addaleax 

> I assume libuv would keep going with the current implementation as its default, right?

Yes.

> In that case I’m not sure how I feel about that – 1845 fixed a real bug, imo…

In my opinion, #1845 was a policy decision: "Let's not let DNS hog the threadpool". This is the "TP Throttling" design, see number 3 in the list of designs in the meta-issue #1959. #1845 was certainly a valuable contribution, don't get me wrong! But if we are to go with a pluggable threadpool in the long run, then shouldn't we leave policy decisions like that to the embedder?

I am of the opinion that we should make the default libuv executor as simple as possible -- a single-producer multi-producer threadpool, as it was before #1845. An embedder that wants anything more sophisticated can define their own executor.

Using the `uv_work_options_t` it will be straightforward to implement #1845 or more sophisticated TP Throttling in an embedder threadpool, or other approaches like a multi-pool along the lines of "TP Handles".

I completely agree that something along the lines of #1845 should be present in the executor that Node.js uses. But it's not clear to me that it's something other embedders need. I don't recall seeing anyone but Node.js folks commenting on the DNS behavior in related issues.

Does that make sense? Or am I taking this pluggability theme too far?

davisjam commented:
This PR now contains a working implementation of the "pluggable threadpool" design.
The tests in 6d5587b include a working userland executor implementation.

See `docs/src/threadpool.rst` for the API.

Some things to note:
- The internal users of the executor (FS, DNS) now call the external `uv_executor_queue_work` API rather than taking an internal route into the threadpool. This cleans up the API for pluggability but requires async FS and DNS calls to `uv__malloc`/`uv__free` a `uv_work_t`. This may have an undesirable performance impact.
- I am using `uv_req_t.reserved[0]` in order to obtain ABI compatibility.
- As I mentioned in previous comments in this PR, I have reverted the default libuv threadpool (executor) to pre- #1845.

TODOs:
- Currently UNIX-only, see commit message ee94bae for the missing Windows pieces.
- Internal users of the executor don't currently provide a `uv_work_options`
- I don't know what the expected behavior on `fork` is, so those tests don't work right now. Please advise.

I will happily clean up the various TODOs, but I want to reach consensus on the general design before I polish up the details.

I believe this PR is ABI compatible.

I cannot figure out how to run the tests under valgrind. Any advice?

@bnoordhuis @addaleax @saghul @jorangreef

davisjam commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,146 @@ Data types
+
+.. c:type:: void (*uv_executor_destroy_func)(uv_executor_t* executor)
+
+    Called at some point ???
```
I put a destroy routine in src/executor.c and wrapped it with UV_DESTRUCTOR. I think this uses atexit hooks?

addaleax replied:
UV_DESTRUCTOR would use the dynamic linker/standard library functions to clean up, i.e. I think it would typically run after `atexit` (and possibly a lot earlier, if libuv is unloaded as a shared library)?

davisjam commented on src/unix/getaddrinfo.c:
```diff
@@ -184,13 +202,18 @@ int uv_getaddrinfo(uv_loop_t* loop,
-                    uv__getaddrinfo_done);
+    /* TODO options should indicate type. */
+    work->data = req;
+    req->reserved[0] = work; /* For uv_cancel. */
```
req is a uv_getaddrinfo_t so req->reserved[0] is private from its perspective. Since req->reserved[0] is part of the executor API I thought it would be OK to use it privately here to give a pointer to the uv_work_t it submits.

This enables uv_cancel.

I do the same thing for uv_getnameinfo and uv_fs.

davisjam commented:
I ran the relevant benchmarks -- `fs_stat` and `getaddrinfo` -- to see the impact of per-request memory de/allocation for FS and DNS operations.

- `fs_stat`: No noticeable difference
- `getaddrinfo`: Hideous performance drop

<img width="512" alt="image" src="https://user-images.githubusercontent.com/22822319/44886699-042e1180-ac96-11e8-873c-22c459cba58a.png">

<img width="512" alt="image" src="https://user-images.githubusercontent.com/22822319/44886700-07c19880-ac96-11e8-9045-aa45f6f14056.png">

davisjam commented:
> Windows FS and DNS are also rewired to use the external executor API. I don't have a Windows machine so I haven't checked compilation or tests.

Could someone with a Windows machine (or CI?) run a Windows job on this branch and let me know about compilation warnings or errors?

davisjam commented:
Bump.

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,156 @@ Data types
+.. c:type:: int (*uv_executor_cancel_func)(uv_executor_t* executor, uv_work_t* req)
+
+    Called when someone wants to cancel a previously-submit'ed request.
+    Return ``UV_EBUSY`` if you cannot cancel it.
```
I think mixing `someone` and `you` is a bit odd?

davisjam replied:
Regarding `uv_executor_destroy_func`: I think the responsibility for initialization and destruction if best left to the embedder. I am planning to remove these unless someone thinks this is a good idea.

davisjam replied:
Done.

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,156 @@ Data types
+
+    Replace the default libuv executor with this user-defined one.
+    Must be called before any work is submitted to the default libuv executor.
+    Returns 0 on successs.
```
typo: `success`

Can we specify the lifetime requirements for `executor` here, and what happens if there is currently executing code? 

Also, is there any chance we could specify executors per-loop at some point? I’d guess that’s tricky because a lot of the code currently isn’t laid out for that?

davisjam replied:
Perhaps I am a snakesss.

> what happens if there is currently executing code?

Once work has been submitted, subsequent calls to `uv_replace_executor` will return non-zero.

> Also, is there any chance we could specify executors per-loop at some point? I’d guess that’s tricky because a lot of the code currently isn’t laid out for that?

You would need to ensure the executor was drained or paused. If paused you would also need a way to transfer work from the old executor to the new one.

Out of curiosity, why would we want this? For online tuning purposes? (If so I think the burden would be on the embedder to figure that out.)

addaleax replied:
I’m partially thinking about tuning purposes, yes, but also about Node’s own embedder scenarios; Node.js is not necessarily the only user of libuv APIs in a process, but we’re most likely going to provide our own executor.

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,156 @@ Data types
 
+.. c:member:: uv_work_cb uv_work_t.work_cb
+
+    Executor should invoke this, not on the event loop.
```
Just checking; It would be *okay* to run this on the same thread as the event loop, and possibly during it, it’s just not the intended way to use this?

davisjam replied:
Permitting embedders to run on the same thread as the event loop might break existing libuv users. Some applications may e.g. assume that code running on the event loop can synchronize (not deadlock) with code running in a `work_cb`.

addaleax replied:
Okay, but that is the embedder’s/user’s problem then, right? I guess my question here is: Do you know if the work requests that libuv itself generates depend on this constraint in some way?

davisjam replied:
> Do you know if the work requests that libuv itself generates depend on this constraint in some way?

About 95% sure it does not.

addaleax commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,156 @@ Data types
+
+.. c:member:: uv_executor_cancel_func uv_executor_t.cancel
+
+    Can be NULL.
```
Would it make sense to use a different return value for `uv_cancel` (e.g. `ENOSYS`) if the executor doesn’t support cancelling?

davisjam replied:
Good idea.

davisjam replied:
Fixed.

addaleax commented on src/executor.c:
```diff
@@ -0,0 +1,246 @@
+static volatile int initialized = 1; /* Protected by once in
+                    uv_executor_queue_work, but not in uv_replace_executor. */
+
+uv_executor_t * uv__executor(void) {
```
Not sure which style is the preferred one here, but I assume we want to have only left-leaning or right-leaning pointers in libuv?

davisjam replied:
Yeah.

davisjam commented:
Update: Addressed actionable suggestions from addaleax.

This has now slightly diverged from my Node.js PR. Will sync these up later.

bnoordhuis commented on src/threadpool.c:
```diff
@@ -29,350 +29,277 @@
-    q = QUEUE_HEAD(&wq);
-    QUEUE_REMOVE(q);
+  fields = (struct default_executor_fields *) executor->data;
+	assert(fields);
```
Tab again.  Also, please do an explicit check: `fields != NULL`

bnoordhuis commented on src/threadpool.c:
```diff
@@ -29,350 +29,277 @@
+  /* Put executor-specific data into req->reserved[0]. */
+  wreq = &req->work_req;
+  req->reserved[0] = wreq;
+  wreq->work = 0xdeadbeef; /* Non-NULL: "Not yet completed". */
```
This isn't legal C if work is a function pointer.  Didn't the compiler warn?

davisjam replied:
No comment.

bnoordhuis commented on src/threadpool.c:
```diff
@@ -29,350 +29,277 @@
+    executor->done(req);
+    return 0;
+  }
+  else {
```
`else` should go on the same line as the `}`.

bnoordhuis commented on src/threadpool.c:
```diff
@@ -29,350 +29,277 @@
+uv_executor_t * uv__default_executor(void) {
+  uv_once(&init_default_executor_once, uv__default_executor_init);
+  return &default_executor;
 }
```
I will admit that I don't really the see the value of grouping everything in that `_fields` struct when it's still all global state.

davisjam replied:
It made a bit more sense when the executor API had `init` and `destroy` callbacks. I think I will revert `_fields` now that those are gone.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -90,8 +97,11 @@
       req->path = uv__strdup(path);                                           \
-      if (req->path == NULL)                                                  \
+      if (req->path == NULL) {                                                \
+        if (work)                                                             \
```
`work != NULL`.  Likewise on line 121, 1426, etc.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -119,12 +132,18 @@
+                      work,                                                   \
+                      &options,                                               \
+                      uv__fs_executor_work,                                   \
+                      uv__fs_executor_done);                                  \
```
Style: arguments should line up.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -1566,11 +1693,16 @@ int uv_fs_copyfile(uv_loop_t* loop,
+      uv__free(work);
     return UV_EINVAL;
   }
 
```
Introducing a memory allocation for each fs request is kind of regrettable.  Do you think there is a way to avoid that?

davisjam replied:
The FS (and DNS) code formerly took an internal route to the threadpool using the hidden `uv__work` member hidden in those requests. Now they must formally submit work to the executor like everyone else.

Thus they need a `uv_work_t`.

One solution to avoid per-request allocation is to pre-allocate some `uv_work_t`'s, statically or during initialization, and maintain a list of free `uv_work_t`'s. If the pre-allocated set runs out then we can allocate new `uv_work_t`'s and either add them to the free list when they are done (up to some high water mark) or free them afterwards.

bnoordhuis commented on src/unix/getaddrinfo.c:
```diff
@@ -153,14 +163,23 @@ int uv_getaddrinfo(uv_loop_t* loop,
     return UV_ENOMEM;
 
+  work = NULL;
+  if (getaddrinfo_cb) {
```
`getaddrinfo_cb != NULL`

bnoordhuis commented on src/win/getaddrinfo.c:
```diff
@@ -311,6 +320,14 @@ int uv_getaddrinfo(uv_loop_t* loop,
 
   /* save alloc_ptr now so we can free if error */
   req->alloc = (void*)alloc_ptr;
+  work = NULL;
```
Do this before the first `goto error` statement.  It's potentially freeing a wild pointer now.

davisjam replied:
Good catch.

bnoordhuis commented on test/test-executor.c:
```diff
@@ -0,0 +1,322 @@
+  int n_completed;
+
+  int head;
+  int tail;
```
Libuv generally uses `unsigned int` (or just `unsigned`) for integers that are always >= 0.

bnoordhuis commented on test/test-executor.c:
```diff
@@ -0,0 +1,322 @@
+  uv_work_t* work;
+  int should_break;
+
+  data = (struct toy_executor_data *) arg;
```
Unnecessary cast.  Happens in a few more places.

bnoordhuis commented on test/test-executor.c:
```diff
@@ -0,0 +1,322 @@
+  data = (struct toy_executor_data *) arg;
+
+  should_break = 0;
+  while (1) {
```
Libuv mostly uses `for (;;)` for this (although not exclusively, so it's not wrong, just less idiomatic.)

bnoordhuis commented on test/test-executor.c:
```diff
@@ -0,0 +1,322 @@
+        break;
+      }
+
+      printf("worker: found %d work -- head %d tail %d\n", data->tail - data->head, data->head, data->tail);
```
Long line.

davisjam replied:
This will be cut when I purge the debugging messages.

bnoordhuis commented on test/test-executor.c:
```diff
@@ -0,0 +1,322 @@
+static void toy_slow_work(uv_work_t* req) {
+  printf("toy_slow_work: req %p\n", req);
+  ASSERT(req);
+  uv_sem_wait((uv_sem_t *) req->data);
```
Use `finish_slow_work` directly here?  That makes it clear what it's waiting on.

bnoordhuis commented on test/test-executor.c:
Long line.

bnoordhuis commented on test/test-fs.c:
```diff
@@ -1259,6 +1259,8 @@ TEST_IMPL(fs_access) {
   unlink("test_file");
   rmdir("test_dir");
 
+  printf("Hello world\n");
```
Stray debug code, I presume? :-)

davisjam commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,139 @@ Data types
+.. c:member:: void* uv_work_t.reserved[0]
+
+    Space for arbitrary data. libuv does not use this field.
+    This is intended for use by an executor implementation.
```
Makes sense, thanks for the tip.

davisjam commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,139 @@ Data types
+    Suggested for use in user-defined executor.
+    Has no effect on libuv's default executor.
+
+.. c:member:: int uv_work_options_t.cancelable
```
Correct, it is not currently used. This would permit the executor to attempt to forcibly abort tasks that are already running on a worker. Since `pthread_cancel` is pretty dangerous, I'm not sure we want to go there, but I wanted to float the idea.

As an alternative I suppose the executor could detect long-running work and spin up replacement workers instead.

If the long-running work is hung on I/O (e.g. a blocking read), pthread_cancel probably won't kill it until the I/O finishes. So this would mostly be helpful for "stuck" CPU-bound work that will continue to burn the CPU until it finishes or it is aborted.

davisjam commented on src/executor.c:
```diff
@@ -0,0 +1,233 @@
+  _executor->done = uv__executor_done;
+
+  /* Replace our executor. */
+  executor = _executor;
```
Acknowledged.

Somehow the executor needs to know how to tell libuv that the work is done.

Reasonable alternatives to updating the struct:

1. Expose the "work is done" function in `uv.h`.
2. Accept a pointer to a function pointer.

davisjam commented on src/executor.c:
```diff
@@ -0,0 +1,233 @@
+static void uv__executor_init(void) {
+  /* Assign executor to default if none was set. */
+  if (executor == NULL) {
+    assert(!uv_replace_executor(uv__default_executor()));
```
> Don't perform actions with side effects inside an assert().

OK

> I can't find where uv__default_executor() is defined

It's defined in `src/threadpool.c` since that is where the default executor is implemented. It basically returns a pointer to the `uv_executor_t default_executor` owned by `src/threadpool.c`.

davisjam commented on src/executor.c:
```diff
@@ -0,0 +1,233 @@
+    return UV_EINVAL;
+
+  /* Register req on loop. */
+  LOG_1("uv_executor_queue_work: req %p\n", (void *) req);
```
Yes.

davisjam commented on src/executor.c:
```diff
@@ -0,0 +1,233 @@
+    LOG_0("uv_executor_queue_work: no options provided\n");
+
+  /* Submit to the executor. */
+  executor->submit(executor, req, opts);
```
> Should submit() be able to return a status code?

In principle, yes.

In v1.x, doing so would conflict with the `uv_queue_work` API which does not give a way to return an error code. Existing users of `uv_queue_work` assume that calls to it succeed, so the executor has to honor this.

davisjam commented on src/threadpool.c:
```diff
@@ -29,350 +29,277 @@
+  struct default_executor_fields* fields;
+
+  executor = ((struct worker_arg*) arg)->executor;
+	assert(executor);
```
My .vimrc is rather unhappy.

davisjam commented on docs/src/threadpool.rst:
```diff
@@ -32,35 +55,139 @@ Data types
+
+    Called when the executor should handle this request.
+
+.. c:type:: int (*uv_executor_cancel_func)(uv_executor_t* executor, uv_work_t* req)
```
I was modelling this on [uv_replace_allocator](http://docs.libuv.org/en/v1.x/misc.html).

davisjam replied:
Should I stick with `uv_replace_allocator` style or use the `_cb` style you requested?

davisjam replied:
Bump.

addaleax replied:
If there is no follow-up here I think marking this as resolved is okay

davisjam commented on docs/src/threadpool.rst:
```diff
@@ -190,4 +178,12 @@ API
+    An executor should invoke this function once it finishes with a request.
+    The effect is to return control over the `req` to libuv.
+
+    This function is thread safe. <-- TODO This seems desirable so the executor workers don't have to centralize returns through the event loop, but thread safety requires locking loop->wq_mutex. I'm having trouble imagining how this could lead to deadlock in a "reasonable" executor implementation, but wanted to discuss.
```
Thoughts on this?

davisjam commented:
@bnoordhuis I've largely addressed your review, broken down into separate commits in case you wanted to double-check any of my changes.

PTAL at 9c92d97 (in particular, [this](https://github.com/libuv/libuv/pull/1726/commits/9c92d97da0b85d9aad435eb630f2da4ce8c4064b#r219368859)).

davisjam commented:
Rebased on v1.x

davisjam commented:
It's been about 2 weeks since the last review. In that time I've fixed up a few TODOs and rebased on `v1.x` for ease of merging.

There are still some open questions on this PR. I am hoping for discussion on:

1. Are we OK with per-FS/DNS-request memory allocation/de-allocation?
    -  If this is a problem I think a good `v1.x` approach is to have `src/executor.c` manage a pre-allocated set of, say, 128 `uv_work_t`s that FS and DNS can draw from. Only allocate when this set is fully used.
2. Should `uv_executor_return_work` be thread-safe?
3. Thoughts on the `uv_work_options_t` members?

And here are the remaining TODOs I have to do:

1. Fix the failing fork test.
2. Remove debugging code (logging).
3. Fix any issues on Windows.

bnoordhuis commented:
I haven't forgotten about this PR (it's been open in my browser for some days) but it's pretty big so I need some time to work through it. Hopefully I'll get to it tomorrow.

> Are we OK with per-FS/DNS-request memory allocation/de-allocation?

The extra allocation is regrettable but I wouldn't optimize it prematurely.  I suspect it doesn't have much impact overall.

davisjam commented:
@addaleax @bnoordhuis Hmm?

davisjam commented:
@addaleax I appreciate the comments so far. Can you let me know your thoughts about the general concept, too? "Is this a thing libuv wants, and is this the approach libuv wants?"

addaleax commented:
@davisjam I would, personally, like to see this feature, yes, and I think the general approach seems good to me as well. :+1: Can’t speak for other libuv maintainers, though.

addaleax commented:
ping @davisjam? :)

davisjam commented:
Paper deadline tomorrow. Want to read about regexes? ;-)

Expect an update next week!

addaleax commented:
@davisjam Sure! :smile:

davisjam commented:
@addaleax My most recent commits addressed your comments.

1. The behavior on fork still needs to be resolved.
2. Remove logging.
3. Confirm I did not break Windows.

addaleax commented:
@davisjam I’ll take a look in more detail later, but I can run CI for you now: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1258/

davisjam commented:
Thanks @addaleax. Looks like:

- Fork failures
- Windows failure due to missing `strings.h`.

I'll push fixes and request another run tomorrow.

davisjam commented:
@addaleax 

1. 3bea76e contains fork support. See the TODO in threadpool.c:uv__default_executor_fork though. Not sure whether we should address this. It looks like we are relying on working-but-unspecified pthread behavior after fork?
2. d403ba2 should fix the Windows build issue.

Can you kick off another CI run please?

davisjam commented:
@addaleax CI run? :-)

addaleax commented:
Sorry, of course!

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1269/

davisjam commented:
@addaleax Fixed another Windows build break. Can you CI again when you have a mo'?

addaleax commented:
Sure!

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1270/
